src/machine.rs

```diff
@@ -1382,8 +1382,31 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for MiriMachine<'mir, 'tcx> {
     ) -> InterpResult<'tcx> {
         // We want this *before* the return value copy, because the return place itself is protected
         // until we do `end_call` here.
-        if let Some(borrow_tracker) = &ecx.machine.borrow_tracker {
-            borrow_tracker.borrow_mut().end_call(&frame.extra);
+        if let Some(global_borrow_tracker) = &ecx.machine.borrow_tracker {
+            // The body of this loop needs `global_borrow_tracker` immutably
+            // so we can't move this code inside the following `end_call`.
+            for (alloc_id, tag) in &frame
+                .extra
+                .borrow_tracker
+                .as_ref()
+                .expect("we should have borrow tracking data")
+                .protected_tags
+            {
+                // If this fails, *do not* return it as an error, it just
+                // means that the allocation no longer exists, and since this is
+                // only going to produce implicit reads and not actual ones, we
+                // don't want that to produce errors.
+                if let Ok(alloc_extra) = ecx.get_alloc_extra(*alloc_id) {
```

Comment:
This was intended as a protection against trying to access an allocation that was freed.
Just because there is a protector does not guarantee that the allocation exists (weak protector). What's the canonical way of checking if an allocation still exists ?

Reply:
You can call `get_alloc_info`.